File: backend/firmsignal/api/validation.py
import re

from fastapi import HTTPException


# Limits
MIN_LENGTH = 2
MAX_LENGTH = 100

# Characters valid in a company name:
# letters, numbers, spaces, dots, commas, ampersands, hyphens, apostrophes, parens
VALID_PATTERN = re.compile(r"^[a-zA-Z0-9\s\.\,\&\-\'\(\)]+$")

# Known garbage inputs to reject immediately
BLOCKED_INPUTS = {
    "test", "asdf", "qwerty", "foo", "bar",
    "null", "none", "undefined", "na", "n/a",
}


def validate_company_name(raw: str) -> str:
    """
    Validates and sanitizes a company name input.
    Returns the cleaned string or raises HTTPException.

    Called before anything else in POST /analyze.
    """

    # Step 1 — strip whitespace
    cleaned = raw.strip()

    # Step 2 — check not empty
    if not cleaned:
        raise HTTPException(
            status_code=422,
            detail="Company name cannot be empty.",
        )

    # Step 3 — check minimum length
    if len(cleaned) < MIN_LENGTH:
        raise HTTPException(
            status_code=422,
            detail=f"Company name must be at least {MIN_LENGTH} characters.",
        )

    # Step 4 — check maximum length
    if len(cleaned) > MAX_LENGTH:
        raise HTTPException(
            status_code=422,
            detail=f"Company name cannot exceed {MAX_LENGTH} characters.",
        )

    # Step 5 — check for valid characters only
    if not VALID_PATTERN.match(cleaned):
        raise HTTPException(
            status_code=422,
            detail=(
                "Company name contains invalid characters. "
                "Only letters, numbers, spaces, and basic punctuation are allowed."
            ),
        )

    # Step 6 — reject obvious garbage
    if cleaned.lower() in BLOCKED_INPUTS:
        raise HTTPException(
            status_code=422,
            detail=f"'{cleaned}' is not a valid company name.",
        )

    # Step 7 — normalize multiple spaces to a single space
    cleaned = re.sub(r"\s+", " ", cleaned)

    return cleaned

File: frontend/lib/validation.ts
export interface ValidationResult {
  valid: boolean
  error: string | null
}

const VALID_PATTERN = /^[a-zA-Z0-9\s.,&\-'()]+$/
const MIN_LENGTH = 2
const MAX_LENGTH = 100
const BLOCKED = new Set(["test", "asdf", "qwerty", "foo", "bar", "null", "none", "undefined", "na", "n/a"])

/**
 * Real-time checks — runs on every keystroke.
 * Only covers errors that are unambiguous mid-input:
 *   - exceeded max length
 *   - invalid character just typed
 * Intentionally skips empty / too-short / blocked-word checks
 * (those are only meaningful after the user signals they're done).
 */
export function getLiveError(value: string): string | null {
  if (value.length > MAX_LENGTH) {
    return `Company name cannot exceed ${MAX_LENGTH} characters`
  }
  if (value.length > 0 && !VALID_PATTERN.test(value)) {
    return "Only letters, numbers, and basic punctuation allowed"
  }
  return null
}

export function validateCompanyInput(raw: string): ValidationResult {
  const cleaned = raw.trim()

  if (!cleaned) {
    return { valid: false, error: "Please enter a company name" }
  }

  if (cleaned.length < MIN_LENGTH) {
    return { valid: false, error: "Company name is too short" }
  }

  if (cleaned.length > MAX_LENGTH) {
    return {
      valid: false,
      error: `Company name cannot exceed ${MAX_LENGTH} characters`,
    }
  }

  if (!VALID_PATTERN.test(cleaned)) {
    return {
      valid: false,
      error: "Only letters, numbers, and basic punctuation allowed",
    }
  }

  if (BLOCKED.has(cleaned.toLowerCase())) {
    return { valid: false, error: `"${cleaned}" is not a valid company name` }
  }

  return { valid: true, error: null }
}
